src/coreclr/inc/corinfo.h

```diff
@@ -2403,6 +2441,42 @@ class ICorStaticInfo
             int32_t                     num
             ) = 0;
 
+    //------------------------------------------------------------------------------
+    // getTypeLayout: Obtain a tree describing the layout of a type.
+    //
+    // Parameters:
+    //   typeHnd            - Handle of the type.
+    //   treeNodes          - [out] Pointer to tree node entries to write.
+    //   numTreeNodes       - [in, out] Size of 'treeNodes'. Updated to contain
+    //                         the number of entries written in 'treeNodes'.
+    //
+    // Returns:
+    //   A result indicating whether the type layout was successfully
+    //   retrieved and whether the result is partial or not.
+    //
+    // Remarks:
+    //   The type layout should be stored in preorder in 'treeNodes': the root
+    //   node is always at index 0, and the first child of any node is at its
+    //   own index + 1.
```

Comment:
No, there is no such assumption. I think the field iteration on the VM side returns fields in metadata order, so with auto type layout and/or ExplicitLayout structs the offsets can be unordered. Regular promotion sorts the fields in some cases because of it, and `StructSegments` in physical promotion does not care about the order.

I will add a comment to explicitly call it out.